datafusion/substrait/src/logical_plan/producer.rs

```diff
@@ -512,6 +572,62 @@ pub fn to_substrait_rel(
     }
 }
 
+fn to_substrait_named_struct(schema: &DFSchemaRef) -> Result<NamedStruct> {
+    // Substrait wants a list of all field names, including nested fields from structs,
+    // also from within lists and maps. However, it does not want the list and map field names
+    // themselves - only structs are considered to have useful names.
+    fn names_dfs(dtype: &DataType) -> Result<Vec<String>> {
+        match dtype {
+            DataType::Struct(fields) => {
+                let mut names = Vec::new();
+                for field in fields {
+                    names.push(field.name().to_string());
+                    names.extend(names_dfs(field.data_type())?);
+                }
+                Ok(names)
+            }
+            DataType::List(l) => names_dfs(l.data_type()),
```

Comment:
I think we also need to consider `DataType::LargeList` here because `to_substrait_type` supports it.

Reply:
Thanks, fixed https://github.com/apache/datafusion/pull/10531/commits/9c83ba469ece074a9611f4db871f2e0044652605

Reply:
`arrow_cast` is a solution .
```sql
VALUES(arrow_cast([1,2], 'LargeList(Int64)'));
```

Reply:
Nice! I added that in https://github.com/apache/datafusion/pull/10531/commits/4b8babe5353cff2740c1211cff6973dc83264098, though it seems that arrow_cast doesn't [yet support structs](https://github.com/apache/datafusion/blob/2e6beb434d45693bc8c7ba636b0407c26c04fc37/datafusion/functions/src/core/arrow_cast.rs#L194), so the test still wouldn't have caught this. Unless you can think of some workaround? I'm not familiar at all with arrow_cast and the like 😅

Reply:
I think so too, let's keep it as a TODO.